ecpy_hqc_legacy/instruments/drivers/dll/alazar935x.py

```diff
@@ -116,399 +62,221 @@ def close_connection(self):
-
-            data = np.reshape(buffer.buffer, (recordsPerBuffer*channel_number, -1))
-            dataA[buffersCompleted*recordsPerBuffer:(buffersCompleted+1)*recordsPerBuffer] = data[:recordsPerBuffer]
-            dataB[buffersCompleted*recordsPerBuffer:(buffersCompleted+1)*recordsPerBuffer] = data[recordsPerBuffer:]
-            buffersCompleted += 1
+            buffer = buffers[buffers_completed % len(buffers)]
+            board.waitAsyncBufferComplete(buffer.addr, timeout_ms=5000)
+            rbuf = np.reshape(buffer.buffer,
+                              (records_per_buffer*channel_count,
+                               samples_per_record))
+            # making sure we only grab the number of records we asked for
+            if buffers_completed < buffers_per_acquisition-1:
+                records_to_ignore_val = 0
+            else:
+                records_to_ignore_val = records_to_ignore
 
-            self._dll.PostAsyncBuffer(board, buffer.addr, buffer.size_bytes)
+            start = buffers_completed*records_per_buffer
+            stop = start + records_per_buffer-records_to_ignore_val
+            for i in range(channel_count):
+                if average:
+                    data[i] += np.sum(rbuf[i*records_per_buffer:
+                                           (i+1)*records_per_buffer -
+                                           records_to_ignore_val], 0)
+                else:
+                    data[i][start:stop] = rbuf[i*records_per_buffer:
+                                               (i+1)*records_per_buffer -
+                                               records_to_ignore_val]
 
-        self._dll.AbortAsyncRead(board)
+            buffers_completed += 1
 
-        for buffer in buffers:
-            buffer.__exit__()
+            # Add the buffer to the end of the list of available buffers.
+            board.postAsyncBuffer(buffer.addr, buffer.size_bytes)
 
-        # Re-shaping of the data for demodulation and demodulation
-        dataA = dataA[:,1:samplesPerTrace + 1]
-        dataB = dataB[:,1:samplesPerTrace + 1]
+        # Abort transfer.
+        board.abortAsyncRead()
 
-        # Averaging if needed and converting binary numbers into Volts
         if average:
-            dataA = np.mean(dataA, axis=0)
-            dataB = np.mean(dataB, axis=0)
-
-        dataA = (dataA-2**15)/65535*0.8+0.000459610322728
-        dataB = (dataB-2**15)/65535*0.8+0.00154325074388
-
-        return (dataA, dataB)
-
+            for c in range(channel_count):
+                data[c] /= records_per_capture
+
+        # Check card is not saturated
+        maxADC = 2**16-100
+        minADC = 100
+        if any(np.max(data[i]) > maxADC or np.min(data[i]) < minADC for i in
+               range(channel_count)):
+            mes = '''Channel A or B are saturated: increase input range or
+            decrease amplification'''
+            raise InstrIOError(mes)
+
+        # XXX convert to volt
```

Comment:
the converstion to volts requires knowledge of the voltage range of the alazar. We could implement this in the future, but it is not really needed. We find it useful to know the ADC value in order to know wheter the card is saturated or not.